### ISSUE TITLE
cordova create does not accept valid domain names containing dashes

### ISSUE THREAD

**Issue by tx350z** [apache/cordova#184]
### Issue Type

- [ x] Bug Report
- [ ] Feature Request
- [ ] Support Question

## Information

Attempted to create an app with an ID for a domain name that contains a dash (-). Create failed with the message "App id contains a reserved word, or is not a valid identifier."

Changed the app id to a different domain and it succeeded. Internet domain names containing dashes have been supported since the beginning of time.

### Command or Code

cordova create hello com.my-domain-name.hello HelloWorld

### Environment, Platform, Device

New install of node.js, npm, and Cordova on Windows 8.1 with all latest updates.

### Version information

cordova cli

## Checklist

- [ x] I searched for already existing GitHub issues about this
- [ x] I updated all Cordova tooling to their most recent version
- [ x] I included all the necessary information above

**Comment by breautek** [comment]
To be clear, this isn't a domain, it is a reverse dns notation. It is used as a way to builld unique ids. In the case of android platforms, it is used to create packages, which hyphens are an illegal character. This is why we cannot accept an app id with hyphens. This is working as intended.

I suggest using a different name for app id, either omitting the hyphens, or replacing them with underscores.

**Comment by purefan** [comment]
- If you say "reverse dns notation" then we expect it to accept hyphens
- If you can't accept hyphens because of Android that's ok, but then we expect it to be documented (I could not find this behavior anywhere but maybe I missed it)

For anyone else coming here through a web search:
This repository pulls dependencies from other apache repositories, the validation for the id happens when this project calls [the cordova-cli routine for create](https://github.com/apache/cordova-cli/blob/838a05784a3bc040fbf1ec223fba976663954234/src/cli.js#L348) which in turn [runs the validation](https://github.com/apache/cordova-create/blob/5f24bdbceba46830a46dc940306688b2d71f447a/index.js#L68) from [valid-identifier](https://github.com/purplecabbage/valid-identifier)

So one way to fix this _situation_ is to improve valid-identifier to throw an Error or implement some mechanism to bubble the offending part, seeing how valid-identifier has not been updated in a few years Id say chances of pull requests being accepted are slim

In other words, maybe the best place for a bug report is the cordova-cli project, and not this one.

**Comment by raphinesse** [comment]
> So one way to fix this _situation_ is to improve valid-identifier to throw an Error or implement some mechanism to bubble the offending part, seeing how valid-identifier has not been updated in a few years Id say chances of pull requests being accepted are slim

Actually, the `valid-identifier` package is maintained by Cordova's current PMC chair, @purplecabbage. So if there would be a quality contribution to improve it, I'm sure it would be considered.

I don't think opening an issue in cordova-cli would help. I rather suggest you open a pull request on our docs to fix the lack of documentation regarding the constraints for app identifiers.

**Comment by breautek** [comment]
> If you say "reverse dns notation" then we expect it to accept hyphens

Yes, this is a "gotcha". Reverse DNS notation is what is commonly used in java/android land. But it's used for package and class names where the hyphen character is illegal.

> If you can't accept hyphens because of Android that's ok, but then we expect it to be documented

I always feel the documentation can be improved and this could be one of those things, especially since Cordova tries to abstract native details away from its users... I raised an issue for this at https://github.com/apache/cordova-docs/issues/1087
